### ISSUE TITLE
Testing: Investigate creation of an EnumVerifier

### ISSUE THREAD

**Issue by ljacqu** [Xephi/AuthMeReloaded#975]
Check if it makes sense to create some sort of EnumVerifier class for our tests. For many enums we typically test the same thing: does each entry have a unique, non-empty code/key?

Unclear whether it's possible to create a class generic enough to handle multiple fields (e.g. PermissionsSystemType) or to handle uniqueness checks over multiple enum types (e.g. PlayerPermission / AdminPermission / PlayerStatePermission).

**Comment by ljacqu** [comment]
I'm resolving this as a **wontfix**. There are several drawbacks, two are technically quite interesting and I'd like to outline them here below.

#### 1. Cannot specify a generic _interface_
An enum verifier class would have been interesting to work with enums that implement an interface such as `PermissionNode`. However, Java generics don't offer a way to specify a generic interface. 

For example, the following is fine, because `T` can be a class or an interface.
```java
public <T extends Enum<T> & PermissionNode> void accept(Class<T>... elems) {
}
```
However, it is not possible to restrict classes to being Enums and implementing some generic interface. Something like:
```java
public class Test<I> {
    public <T extends Enum<T> & I> void accept(Class<T>... elems) { // oops!
    }
}
```
because `I` can't be restricted to interfaces (it could be a class), and Java generics only accept one class as `extends` bound.

#### 2. Runtime exception is thrown when generic type isn't given explicitly
The best way to circumvent the above problem is to just initialize with an `I` type and to accept a `Class<? extends I>...` parameter and to then check manually that all provided class are enums.

The idea once that this is done is that you can pass some lambdas taking/producing `I` elements to execute some sort of test across all enum entries. However, if you don't explicitly specify the interface as the class' generic type, an exception is thrown at runtime:

> java.lang.BootstrapMethodError: call site initialization exception
Caused by: java.lang.invoke.LambdaConversionException: Invalid receiver type class java.lang.Enum; not a subtype of implementation type interface {MyInterface}

Here's a minimal, runnable example that will throw a runtime exception (as tested with Java 1.8.0_101). If you do as the comment in `main()` says, the error disappears.

```java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

/** Causes a runtime exception. */
public class EnumProcessor<I> {

    private List<I> elements = new ArrayList<>();

    @SafeVarargs
    private EnumProcessor(Class<? extends I>... classes) {
        Arrays.stream(classes)
            .peek(clz -> {
                if (!clz.isEnum()) throw new IllegalArgumentException(clz.getName());
            })
            .map(Class::getEnumConstants)
            .forEach(elems -> elements.addAll(Arrays.asList(elems)));
    }

    @SafeVarargs
    public static <I> EnumProcessor<I> create(Class<? extends I>... classes) {
        return new EnumProcessor<>();
    }

    public <R> List<R> transform(Function<I, R> transformer) {
        return elements.stream().map(transformer).collect(Collectors.toList());
    }

    public static void main(String... args) {
        // Change EnumProcess.create to EnumProcess.<MyInterface>create and it works
        EnumProcessor.create(MyEnum.class, MyOtherEnum.class)
            .transform(MyInterface::getKey);
    }

    private interface MyInterface {
        String getKey();
    }

    private enum MyEnum implements MyInterface {
        FIRST, SECOND, THIRD;

        @Override
        public String getKey() {
            return "test";
        }
    }

    private enum MyOtherEnum implements MyInterface {
        FOURTH, FIFTH;

        @Override
        public String getKey() {
            return "1234";
        }
    }
}
```
